_posts/2022-04-01-tidying-snapshots-up.md

```diff
@@ -0,0 +1,213 @@
+and thus not being able to clean those files up. To solve this issue we added explicit restore
+modes:
+CLAIM, NO_CLAIM and LEGACY (for backwards compatibility) which clearly define if Flink should take
+care of cleaning up the snapshot or should it remain in users responsibility.
+
+# Restore mode
+
+The `Restore Mode` determines who takes ownership of the files that make up a savepoint or an
+externalized checkpoints after restoring it. Snapshots, which are either checkpoints or savepoints
+in this context, can be owned either by a user or Flink itself. If a snapshot is owned by a user,
+Flink will not delete its files. What is even more important, Flink can not depend on the existence
+of the files from such a snapshot, as it might be deleted outside of Flink's control.
+
+To begin with let's see how did it look so far and what problems it may pose. We left the old
+behaviour which is available if you pick the `LEGACY` mode.
+
+## LEGACY mode
+
+The legacy is mode is how Flink worked until 1.15. In this mode Flink will never delete the initial
+checkpoint. Unfortunately, at the same time, it is not clear if a user can ever delete it as well. 
+The problem here, is that Flink might immediately build an incremental checkpoint on top of the
+restored one. Therefore, subsequent checkpoints depend on the restored checkpoint. Overall, the
+ownership is not well-defined in this mode.
+
+<div style="text-align: center">
+  <img src="{{ site.baseurl }}/img/blog/2022-04-xx-tidying-snapshots/restore-mode-legacy.svg" alt="LEGACY restore mode" width="70%">
+</div>
+
+To fix the issue of a files that no one can reliably claim ownership we introduced the `NO_CLAIM`
+mode as the new default.
+
+## NO_CLAIM (default) mode
+
+In the *NO_CLAIM* mode Flink will not assume ownership of the snapshot. It will leave the files in
+user's control and never delete any of the files. In this mode you can start multiple jobs from the
+same snapshot.
+
+In order to make sure Flink does not depend on any of the files from that snapshot, it will force
+the first (successful) checkpoint to be a full checkpoint as opposed to an incremental one. This
+only makes a difference for `state.backend: rocksdb`, because all other state backends always take
+full checkpoints.
+
+Once the first full checkpoint completes, all subsequent checkpoints will be taken as
+usual/configured. Consequently, once a checkpoint succeeds you can manually delete the original
+snapshot. You can not do this earlier, because without any completed checkpoints Flink will - upon
+failure - try to recover from the initial snapshot.
+
+<div style="text-align: center">
+  <img src="{{ site.baseurl }}/img/blog/2022-04-xx-tidying-snapshots/restore-mode-no_claim.svg" alt="NO_CLAIM restore mode" width="70%" >
+</div>
+
+If you do not want to sacrifice any performance while taking the first checkpoint, we suggest
+looking into the `CLAIM` mode.
```

Comment:
```suggestion
### NO_CLAIM (default) mode

To fix the issue of files that no one can reliably claim ownership of, we introduced the `NO_CLAIM`
mode as the new default. In this mode, Flink will not assume ownership of the snapshot and will leave the files in
the user's control and never delete any of the files.  You can start multiple jobs from the
same snapshot in this mode.

In order to make sure Flink does not depend on any of the files from that snapshot, it will force
the first (successful) checkpoint to be a full checkpoint as opposed to an incremental one. This
only makes a difference for `state.backend: rocksdb`, because all other state backends always take
full checkpoints.

Once the first full checkpoint completes, all subsequent checkpoints will be taken as
usual/configured. Consequently, once a checkpoint succeeds, you can manually delete the original
snapshot. You can not do this earlier, because without any completed checkpoints, Flink will - upon
failure - try to recover from the initial snapshot.

<div style="text-align: center">
  <img src="{{ site.baseurl }}/img/blog/2022-04-xx-tidying-snapshots/restore-mode-no_claim.svg" alt="NO_CLAIM restore mode" width="70%" >
</div>
```